private/pkg/manifest/module.go

```diff
@@ -0,0 +1,321 @@
+	if hash == nil {
+		return nil, fmt.Errorf("nil hash")
+	}
+	dType, ok := hashKindToDigestType[hash.Kind]
+	if !ok {
+		return nil, fmt.Errorf("unsupported hash kind: %s", hash.Kind.String())
+	}
+	return NewDigestFromBytes(dType, hash.Digest)
+}
+
+// NewMemoryBlobFromReader creates a memory blob from content, which is read
+// until completion. The returned blob contains all bytes read. If you are using
+// this in a loop, you might better use NewMemoryBlobFromReaderWithDigester so
+// you can reuse your digester.
+func NewMemoryBlobFromReader(content io.Reader) (Blob, error) {
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	return NewMemoryBlobFromReaderWithDigester(content, digester)
+}
+
+// NewMemoryBlobFromReaderWithDigester creates a memory blob from content with
+// the passed digester. The content is read until completion. The returned blob
+// contains all bytes read.
+func NewMemoryBlobFromReaderWithDigester(content io.Reader, digester Digester) (Blob, error) {
+	var contentInMemory bytes.Buffer
+	tee := io.TeeReader(content, &contentInMemory)
+	digest, err := digester.Digest(tee)
+	if err != nil {
+		return nil, err
+	}
+	return &memoryBlob{
+		digest:  *digest,
+		content: contentInMemory.Bytes(),
+	}, nil
+}
+
+// BlobEqual returns true if blob a is the same as blob b. The digest is
+// checked for equality and the content bytes compared.
+//
+// An error is returned if an unexpected I/O error occurred when opening,
+// reading, or closing either blob.
+func BlobEqual(ctx context.Context, a, b Blob) (_ bool, retErr error) {
+	const blockSize = 4098
+	if !a.Digest().Equal(*b.Digest()) {
+		// digests don't match
+		return false, nil
+	}
+	aFile, err := a.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, aFile.Close()) }()
+	bFile, err := b.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, bFile.Close()) }()
+	// Read blockSize from a, then from b, and compare.
+	aBlock := make([]byte, blockSize)
+	bBlock := make([]byte, blockSize)
+	for {
+		aN, aErr := aFile.Read(aBlock)
+		bN, bErr := io.ReadAtLeast(bFile, bBlock[:aN], aN) // exactly aN bytes
+		// We're running unexpected error processing (not EOF) before comparing
+		// bytes because it doesn't matter if the returned bytes match if an
+		// error occurred before an expected EOF.
+		if bErr == io.ErrUnexpectedEOF {
+			// b is shorter; we can error early
+			return false, nil
+		}
+		if aErr != nil && aErr != io.EOF {
+			// unexpected read error
+			return false, aErr
+		}
+		if bErr != nil && bErr != io.EOF {
+			// unexpected read error
+			return false, bErr
+		}
+		if !bytes.Equal(aBlock[:aN], bBlock[:bN]) {
+			// Read content doesn't match. Don't forget to forward any errors
+			// returned.
+			return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
```

Comment:
seems to me here that `multierr.Append(nilEOF(aErr), nilEOF(bErr))` won't ever append any err? If EOF is ignored, and if other, it should have been returned in the previous 2 conditionals?

Reply:
I think you're right! I had this block in a different location before. I must have forgot about this when I moved it to where it is now.

```suggestion
			// Read content doesn't match.
			return false, nil
```